File: Namrata-Sanger/Interval List Intersections.js
/*
Day 9
986. Interval List Intersections
Link : https://leetcode.com/problems/interval-list-intersections/
Level : Medium
*/

/**
 * @param {number[][]} firstList
 * @param {number[][]} secondList
 * @return {number[][]}
 */
var intervalIntersection = function(firstList, secondList) {
    let i=0,j=0;
    let intersection = [];
    while(i<firstList.length && j<secondList.length){
        let s1 = firstList[i][0];
        let s2 = secondList[j][0];
        let e1 = firstList[i][1];
        let e2 = secondList[j][1];
        
        if(s2 < e1 && s1 < e2){
            intersection.push([Math.max(s1,s2),Math.min(e1,e2)]);
        }else if(s1 == e2 && e1 > e2){
            intersection.push([s1,s1]);
        }else if(s2 == e1 && e1 < e2){
            intersection.push([e1,e1]);
        } 
        e1 < e2 ? i++ : j++;       
    }
    return intersection;
};

File: Mahesh_Babar/interval_list_intersections.java
// https://leetcode.com/problems/interval-list-intersections/

class interval_list_intersections {
    public int[][] intervalIntersection(int[][] l1, int[][] l2) {
        ArrayList<int[]> ans = new ArrayList<>();
        int n = l1.length, m = l2.length, i = 0, j = 0;
        if (n == 0 || m == 0)
            return ans.toArray(new int[ans.size()][]);
        while (i < n && j < m) {
            int l = Math.max(l1[i][0], l2[j][0]);
            int r = Math.min(l1[i][1], l2[j][1]);
            if (l <= r)
                ans.add(new int[] { l, r });
            if (l1[i][1] < l2[j][1])
                i++;
            else
                j++;
        }
        return ans.toArray(new int[ans.size()][]);
    }
}
